Automattic/woocommerce-payments
# Ensure gateway functionality for disabled payment method

Closes https://github.com/Automattic/woocommerce-payments/issues/8058

#### Problem background
According to https://github.com/Automattic/woocommerce-payments/issues/8058, transactions made with payment methods that were disabled when trying to refund cannot be refunded. This issue arises because WooPayments only registers payment gateways enabled for checkout, whereas WooCommerce Core requires all gateways, including those not enabled for checkout, to be registered. This impacts refunds, among other things. 

Fortunately, there is an `enabled` property in the gateway class that allows a gateway to be enabled or disabled at checkout while remaining registered in the system for other uses, such as refunds. Currently, WooPayments does not utilize this property, and this PR aims to address that.


To address the issue, we need to:

1. Register all payment gateways in WooCommerce Core, not just the ones that are enabled at checkout.
2. Ensure the enabled property accurately indicates whether a gateway is active and can be used at checkout.

Step 2 can be tackled in various ways. The simplest solution was chosen, prioritizing ease of maintenance and minimizing risks. Below is a short analysis of two options, which is optional for reading but can be helpful in case you're curious. 2️⃣ was eventually chosen.

To ensure the enabled property accurately indicates whether a gateway is active we could: 

1️⃣ Re-visit the structure of the `woocommerce_woocommerce_payments_settings` option in the database, which currently has the list of `upe_enabled_payment_method_ids` and the single `enabled` property. The single settings object is used by all gateways, meaning that the `enabled` property can't cover values for separate gateways. There are multiple options to improve this (for example, by replacing `upe_enabled_payment_method_ids` with `payment_methods` and adding the enabled value as a key for every payment method or using separate settings object for every gateway) but all options will require two more things: modify the reading part to accommodate the new structure and implement migration scripts to ensure stores don't lose the state of enabled gateways after the update. 

2️⃣ Leverage the existing `upe_enabled_payment_method_ids` array in the settings, and, while reading from settings (during the gateway initialization step), decide if the gateway is enabled by checking if enabled IDs list contains the gateway's ID. The advantages of such an approach are it's less complex due to way fewer changes, it doesn't require any migration, and it doesn't add complexity to using `wp_options` in WooPayments.

#### Testing instructions

1. Place order with any payment method except card (because it cannot be disabled for testing purposes)
2. After the successful order, navigate to payment settings and disable the payment method used in (1)
3. Navigate to order details -> click payment intent details and you will be redirected to transaction details page
4. In the top right corner, click `Refund in full`, confirm no errors appeared and the payment was refunded, navigate to Stripe Dashboard and confirm the refund was captured there as well
5. Perform steps 1-4, but this time click `Partial refund` instead of `Refund in full`. Partial refund will navigate you to the order details page.
6. After successful step 5, perform steps 1-4 but attempt the full refund from the order page.
7. Access `WooCommerce -> Settings -> Payments`, and toggle the WooPayments option. Save your changes. Refresh the page and confirm the setting is applied correctly, showing it as enabled or disabled.



-------------------


**Post merge**
